solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java

```diff
@@ -244,13 +235,28 @@ private static boolean existCoreName(String coreName, Slice slice) {
         Collections.shuffle(nodeList, random);
       }
     } else {
-      nodeList = new ArrayList<>(liveNodes);
+      nodeList = new ArrayList<>(filterNonDataNodes(zk, liveNodes));
       Collections.shuffle(nodeList, random);
     }
 
     return nodeList;
   }
 
+  public static Collection<String> filterNonDataNodes(DistribStateManager zk, Collection<String> liveNodes) {
```

Comment:
Should this be a generic filter method that takes a Role parameter?

Reply:
I don't see any other use here for filtering on a generic role. In case we need later, we can add so.